File: topological_sort/go/topological_sort.go
package topological_sort

type vertexCallback func(x int)

func Sort(g Graph) []int {
	if !g.Directed {
		panic("Topological sorting only supports directed graphs")
	}

	a := NewAdjacencyList(g)
	sorted := []int{}
	discovered := make([]bool, len(a))

	// Vertices are processed in reverse topological order. We could prepend vertices, but it's faster to append and
	// then reverse at the end.
	process := func(x int) {
		sorted = append(sorted, x)
	}

	for i := range a {
		a.DepthFirstSearch(i, process, discovered)
	}

	reverse(sorted)
	return sorted
}

func reverse(a []int) {
	for i := 0; i < len(a)/2; i++ {
		j := len(a) - 1 - i
		a[i], a[j] = a[j], a[i]
	}
}

File: topological_sort/typescript/topological_sort.ts
import {Graph, AdjacencyList} from './adjacency_list';

export default function sort(graph: Graph): number[] {
    if (!graph.directed) {
        throw new Error('Topological sorting only supports directed graphs.');
    }

    const list = new AdjacencyList(graph);
    const sorted: number[] = [];

    // Reuse the discovered array to avoid processing the same vertices multiple times.
    const discovered: boolean[] = new Array(list.a.length);

    for (let x = 0; x < list.a.length; x += 1) {
        // Vertices are processed in reverse topological order. We could prepend vertices by using `sorted.unshift(x)`,
        // but it's faster to append and then reverse at the end.
        list.depthFirstSearch(x, y => sorted.push(y), discovered);
    }

    sorted.reverse();
    return sorted;
}
